devito/operator.py

```diff
@@ -121,89 +122,62 @@ def __init__(self, expressions, **kwargs):
-                if dim.size is not None:
-                    if not shape[i] <= dim.size:
-                        error('Size of data argument for %s is greater than the size '
-                              'of dimension %s: %d' % (fname, dim.name, dim.size))
-                        raise InvalidArgument('Wrong data shape encountered')
-                    else:
-                        continue
-
-                if dim_sizes[dim.name] is None:
-                    # We haven't determined the size of this dimension yet,
-                    # try to infer it from the data shape.
-                    dim_sizes[dim.name] = shape[i]
-                else:
-                    # We know the dimension size, check if data shape agrees
-                    if not dim_sizes[dim.name] <= shape[i]:
-                        error('Size of dimension %s was determined to be %d, '
-                              'but data for symbol %s has shape %d.'
-                              % (dim.name, dim_sizes[dim.name], fname, shape[i]))
-                        raise InvalidArgument('Wrong data shape encountered')
-
-        # Make sure we have defined all buffered dimensions and their parents,
-        # even if they are not explicitly given or used.
-        d_args = [d for d in arguments.values() if isinstance(d, Dimension)]
-        for d in d_args:
-            if d.is_Buffered:
-                if dim_sizes[d.parent.name] is None:
-                    dim_sizes[d.parent.name] = dim_sizes[d.name]
-                if dim_sizes[d.name] is None:
-                    dim_sizes[d.name] = dim_sizes[d.parent.name]
+        for x in self.t_args + self.dims + self.s_args:
+            x.reset()
+
+    def arguments(self, *args, **kwargs):
+
+        new_params = {}
+        for k, v in kwargs.items():
+            if isinstance(v, CompositeData):
+                orig_param = [x for x in self.symbolic_data if x.name == k][0]
+                for orig_child, new_child in zip(orig_param.children, v.children):
+                    new_params[orig_child.name] = new_child
+        kwargs.update(new_params)
+
+        for ta in self.t_args:
+            assert(ta.verify(kwargs.pop(ta.name, None)))
+
+        for d in self.dims:
+            d.verify(kwargs.pop(d.name, None))
+
+        for s in self.s_args:
+            s.verify(kwargs.pop(s.name, None))
+
+        dim_sizes = OrderedDict([(d.name, d.value) for d in self.dims])
+        dim_names = OrderedDict([(d.name, d) for d in self.dims])
+        dle_arguments = self._dle_arguments(dim_sizes)
+        dim_sizes.update(dle_arguments)
 
+        for d, v in dim_sizes.items():
+            assert(dim_names[d].verify(v))
+        arguments = OrderedDict([(x.name, x.value) for x in self.parameters])
+        arguments.update(self._extra_arguments())
+
+        # Clear the temp values we stored in the arg objects since we've pulled them out
+        # into the OrderedDict object above
```

Comment:
I don't see why that would happen. We just store pointers to all the arguments. If some of the argument objects haven't been allocated already, they need to be allocated before we can get a pointer. However, I don't see why that would happen multiple times.